File: proto/test/Test/Tenants.hs
{-# LANGUAGE OverloadedStrings #-}
-- | Tier 2, the Phase 0 acceptance (docs/ROADMAP.md section 9): every tenant
-- expresses and checks clean; every negative case refuses with exactly its
-- code; the negative goldens cover exactly the codes the checker emits; and
-- the claims section 8 makes about each verdict hold as fields, not only as
-- prose bytes. Also the source-as-data half: every case directory carries
-- its @.rue@ text (unparsed in Phase 0) and every tenant an inventory.
module Test.Tenants (tests) where

import Control.Monad (forM_)
import Data.List (nub, sort)
import qualified Data.Text as T
import Rue.Proto.Check (check)
import Rue.Proto.Intent (Intent (..))
import Rue.Proto.Diagnostics (codeText)
import Rue.Proto.Golden (repoRoot)
import Rue.Proto.Model
import Rue.Proto.Tenants (negatives, tenants)
import Rue.Proto.Tenants.Common
import qualified Rue.Proto.Tenants.T1 as T1
import qualified Rue.Proto.Tenants.T2 as T2
import qualified Rue.Proto.Tenants.T3 as T3
import qualified Rue.Proto.Tenants.T4 as T4
import Rue.Proto.Verdict
import System.Directory (doesFileExist)
import System.FilePath ((</>))
import Test.Check (emittedCodes)
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (assertBool, assertFailure, testCase, (@?=))

tests :: IO TestTree
tests = do
  root <- repoRoot
  pure $
    testGroup
      "tenants"
      [ testGroup "every tenant checks clean" [testCase (T.unpack (tenantName t <> "/" <> caseHost c)) (clean t c) | t <- tenants, c <- tenantCases t]
      , testGroup "every negative refuses with exactly its code" [testCase (negDir n) (refusesWith n) | n <- negatives]
      , testCase "the negative goldens cover exactly the emitted codes" $
          sort (nub (map negCode negatives)) @?= emittedCodes
      , testCase "negative directories are distinct" $
          let ds = map negDir negatives in sort (nub ds) @?= sort ds
      , testCase "case directories are distinct within a tenant" $
          forM_ tenants $ \t -> let hs = map caseHost (tenantCases t) in sort (nub hs) @?= sort hs
      , testGroup
          "the .rue text and inventory exist for every case"
          ( [testCase (T.unpack (tenantName t) <> "/" <> f) (exists (root </> "tenants" </> T.unpack (tenantName t) </> f)) | t <- tenants, f <- ["plan.rue", "inventory.toml"]]
              <> [testCase (negDir n <> "/plan.rue") (exists (root </> "tenants" </> "_negative" </> negDir n </> "plan.rue")) | n <- negatives]
          )
      , testGroup
          "T1 break-glass (section 8.1)"
          [ testCase "temporary, reverts at wane 4h, fully reversible" $ do
              vIntent t1 @?= Just Temporary
              vWane t1 @?= Just (Duration 14400)
              vReversibleThrough t1 @?= 4
              vPointOfNoReturn t1 @?= Nothing
          , testCase "backstop covers 1-2 on the target, installed before 1, armed after 2" $ do
              fmap bvCovers (vBackstop t1) @?= Just [1, 2]
              fmap bvInstalledBefore (vBackstop t1) @?= Just (Just 1)
              fmap bvArmedAfter (vBackstop t1) @?= Just (Just 2)
              fmap bvLateArmingWindow (vBackstop t1) @?= Just [1, 2]
          , testCase "steps 3-4 revert only while the engine lives" $ vControllerOnlyUndos t1 @?= [3, 4]
          , testCase "step 3 is on an API host with controller-side markers" $
              lookup 3 (vHostsTouched t1) @?= Just [HostTouched "bmc-01" "controller"]
          , testCase "the region step's fallback is conditional on no foreign region (task 8)" $ do
              map svConditional (vSteps t1) @?= [Nothing, Just "foreign region in file:/root/.ssh/authorized_keys", Nothing, Nothing]
              fmap (map fst . bvConditional) (vBackstop t1) @?= Just [2]
          , testCase "gate satisfiable by one human, windowed" $ do
              fmap gvSatisfiable (vGate t1) @?= Just True
              fmap gvMinDistinctHumans (vGate t1) @?= Just (Just 1)
              fmap gvWindow (vGate t1) @?= Just (Just (Duration 1800))
          ]
      , testGroup
          "T2 succession (section 8.2)"
          [ testCase "permanent, commits after the guests are up" $ do
              vIntent t2a @?= Just Permanent
              vCommitStep t2a @?= Just 9
              vCommitStep t2m @?= Just 10
          , testCase "reversible through the probes rung; the fence is the point of no return" $ do
              vReversibleThrough t2a @?= 3
              fmap ponrStep (vPointOfNoReturn t2a) @?= Just 4
              fmap ponrGuard (vPointOfNoReturn t2a) @?= Just (Just "fence_verified_off")
              fmap ponrCost (vPointOfNoReturn t2a) @?= Just "fence_verdict"
          , testCase "acknowledged by :none under auto and one human under manual" $ do
              fmap ponrAck (vPointOfNoReturn t2a) @?= Just "none"
              fmap ponrAck (vPointOfNoReturn t2m) @?= Just "humans()"
          , testCase "post-fence holds are held indefinitely; the heir is deferred with a handoff" $ do
              assertBool "no holds" (not (null (vHoldsAt t2a)))
              vHeldIndefinitely t2a @?= [5, 6, 8]
              vHeldIndefinitely t2m @?= [6, 7, 9]
              vDeferred t2a @?= [8]
              vDeferred t2m @?= [9]
          , testCase "the second knell is reachable only on the manual path" $ do
              length [() | sv <- vSteps t2a, svRefusal sv == "knell"] @?= 1
              length [() | sv <- vSteps t2m, svRefusal sv == "knell"] @?= 2
          , testCase "the per-guest loop checks clean under strict" $ do
              planStrictness T2.promoteAuto @?= Strict
              vMayConflicts t2a @?= []
          , testCase "the auto plan contains no human wait" $ do
              vMode t2a @?= "auto"
              vMode t2m @?= "manual"
              vGate t2a @?= Nothing
              map svGate (vSteps t2a) @?= replicate (length (vSteps t2a)) Nothing
              vInducedDefer t2a @?= [5, 6, 8]
          ]
      , testGroup
          "T3 commit-confirmed change (section 8.3)"
          [ testCase "permanent, commits at 4, reversible through 1" $ forM_ [t3, t3w] $ \v -> do
              vIntent v @?= Just Permanent
              vCommitStep v @?= Just 4
              vReversibleThrough v @?= 1
          , testCase "backstop covers step 1, installed and armed before it" $ forM_ [t3, t3w] $ \v -> do
              fmap bvCovers (vBackstop v) @?= Just [1]
              fmap bvInstalledBefore (vBackstop v) @?= Just (Just 1)
              fmap bvArmedBefore (vBackstop v) @?= Just (Just 1)
              fmap bvArmedAfter (vBackstop v) @?= Just Nothing
          , testCase "the region variant's revert is conditional; the Windows owned rule's is not" $ do
              map svConditional (vSteps t3) @?= [Just "foreign region in file:/etc/pf.conf", Nothing, Nothing, Nothing]
              map svConditional (vSteps t3w) @?= replicate 4 Nothing
          ]
      , testGroup
          "T4 reactive host (section 8.4)"
          [ testCase "temporary, reverts at wane, reversible through 1, no backstop" $ forM_ [t4, t4d] $ \v -> do
              vIntent v @?= Just Temporary
              vWane v @?= Just (Duration 7200)
              vReversibleThrough v @?= 1
              vBackstop v @?= Nothing
          , testCase "drift policy stated per step, clobber then defer" $ do
              map svDrift (vSteps t4) @?= [Just "clobber"]
              map svDrift (vSteps t4d) @?= [Just "defer"]
          , testCase "no instance directory: markers on the controller, undo only while the engine lives" $ do
              lookup 1 (vHostsTouched t4) @?= Just [HostTouched "site-ctl" "controller"]
              vControllerOnlyUndos t4 @?= [1]
          ]
      ]
  where
    verdict t c = check (tenantSite t) (tenantRequester t) (casePlan c)
    clean t c = do
      let v = verdict t c
      vDiagnostics v @?= []
      vStatus v @?= Ok
    refusesWith n = do
      let v = check (negSite n) (negRequester n) (negPlan n)
      vStatus v @?= RefusedStatus
      map diagCode (vDiagnostics v) @?= [negCode n]
    negDir n = T.unpack (codeText (negCode n) <> "-" <> negSlug n)
    exists path = do
      ok <- doesFileExist path
      if ok then pure () else assertFailure ("missing " <> path)
    t1 = check T1.site (tenantRequester T1.tenant) T1.breakglass
    t2a = check T2.site (tenantRequester T2.tenant) T2.promoteAuto
    t2m = check T2.site (tenantRequester T2.tenant) T2.promoteManual
    t3 = check T3.site (tenantRequester T3.tenant) T3.openMgmtPort
    t3w = check T3.site (tenantRequester T3.tenant) T3.openMgmtPortWindows
    t4 = check T4.site (tenantRequester T4.tenant) T4.shedLoad
    t4d = check T4.site (tenantRequester T4.tenant) T4.shedLoadDeferring

File: tenants/harness/tests/tenants.rs
//! Tier 2, the Phase 0 acceptance (docs/ROADMAP.md section 9) held by the
//! Rust crates: every tenant checks clean; every negative refuses with
//! exactly its code; the negative goldens cover exactly the codes the checker
//! emits; and the claims section 8 makes about each verdict hold as fields.
//! Also the source-as-data half: every case directory carries its `.rue`
//! text and every tenant an inventory.

use std::collections::BTreeSet;
use std::path::Path;

use rue_core::check::check;
use rue_core::diagnostics::Code;
use rue_core::intent::Intent;
use rue_core::model::{Duration, Strictness};
use rue_core::verdict::{HostTouched, Status, Verdict};
use rue_tenants::golden::repo_root;
use rue_tenants::{load, TenantCase, EMITTED_CODES, NEGATIVES, TENANT_CASES};

fn verdict(root: &Path, input: &str) -> (Verdict, rue_core::ir::PlanIr) {
    let ir = load(root, input).unwrap_or_else(|e| panic!("{e}"));
    (check(&ir.site, &ir.requester, &ir.plan), ir)
}

fn case(root: &Path, tenant: &str, host: &str) -> Verdict {
    verdict(
        root,
        &TenantCase {
            tenant: Box::leak(tenant.to_string().into_boxed_str()),
            host: Box::leak(host.to_string().into_boxed_str()),
        }
        .input(),
    )
    .0
}

#[test]
fn every_tenant_checks_clean() {
    let root = repo_root().unwrap();
    for c in TENANT_CASES {
        let (v, _) = verdict(&root, &c.input());
        assert_eq!(v.diagnostics, vec![], "{}", c.dir());
        assert_eq!(v.status, Status::Ok, "{}", c.dir());
    }
}

#[test]
fn every_negative_refuses_with_exactly_its_code() {
    let root = repo_root().unwrap();
    for n in NEGATIVES {
        let (v, _) = verdict(&root, &n.input());
        assert_eq!(v.status, Status::Refused, "{}", n.name());
        let codes: Vec<Code> = v.diagnostics.iter().map(|d| d.code).collect();
        assert_eq!(codes, vec![n.code], "{}", n.name());
    }
}

#[test]
fn the_negative_goldens_cover_exactly_the_emitted_codes() {
    let covered: BTreeSet<Code> = NEGATIVES.iter().map(|n| n.code).collect();
    let emitted: BTreeSet<Code> = EMITTED_CODES.iter().copied().collect();
    assert_eq!(covered, emitted);
    let mut sorted = EMITTED_CODES.to_vec();
    sorted.sort();
    sorted.dedup();
    assert_eq!(
        sorted, EMITTED_CODES,
        "the emitted list is sorted and duplicate-free"
    );
}

#[test]
fn case_directories_are_distinct() {
    let names: BTreeSet<String> = NEGATIVES.iter().map(|n| n.name()).collect();
    assert_eq!(names.len(), NEGATIVES.len());
    let cases: BTreeSet<String> = TENANT_CASES.iter().map(|c| c.dir()).collect();
    assert_eq!(cases.len(), TENANT_CASES.len());
}

#[test]
fn the_rue_text_and_inventory_exist_for_every_case() {
    let root = repo_root().unwrap();
    let tenants: BTreeSet<&str> = TENANT_CASES.iter().map(|c| c.tenant).collect();
    for t in tenants {
        for f in ["plan.rue", "inventory.toml"] {
            let p = root.join("tenants").join(t).join(f);
            assert!(p.is_file(), "missing {}", p.display());
        }
    }
    for n in NEGATIVES {
        let p = root
            .join("tenants")
            .join("_negative")
            .join(n.name())
            .join("plan.rue");
        assert!(p.is_file(), "missing {}", p.display());
    }
}

#[test]
fn t1_break_glass_section_8_1() {
    let root = repo_root().unwrap();
    let t1 = case(&root, "t1", "db-01");
    assert_eq!(t1.intent, Some(Intent::Temporary));
    assert_eq!(t1.wane, Some(Duration::new(14_400)));
    assert_eq!(t1.reversible_through, 4);
    assert_eq!(t1.point_of_no_return, None);
    let b = t1.backstop.as_ref().unwrap();
    assert_eq!(b.covers, vec![1, 2]);
    assert_eq!(b.installed_before, Some(1));
    assert_eq!(b.armed_after, Some(2));
    assert_eq!(b.late_arming_window, vec![1, 2]);
    assert_eq!(t1.controller_only_undos, vec![3, 4]);
    assert_eq!(
        t1.hosts_touched
            .iter()
            .find(|(n, _)| *n == 3)
            .map(|(_, h)| h.clone()),
        Some(vec![HostTouched::Host {
            host: "bmc-01".into(),
            directory: "controller".into()
        }])
    );
    let conditionals: Vec<Option<String>> =
        t1.steps.iter().map(|s| s.conditional.clone()).collect();
    assert_eq!(
        conditionals,
        vec![
            None,
            Some("foreign region in file:/root/.ssh/authorized_keys".into()),
            None,
            None
        ]
    );
    assert_eq!(
        b.conditional.iter().map(|(n, _)| *n).collect::<Vec<_>>(),
        vec![2]
    );
    let g = t1.gate.as_ref().unwrap();
    assert!(g.satisfiable);
    assert_eq!(g.min_distinct_humans, Some(1));
    assert_eq!(g.window, Some(Duration::new(1800)));
}

#[test]
fn t2_succession_section_8_2() {
    let root = repo_root().unwrap();
    let (auto, ir_auto) = verdict(
        &root,
        &TenantCase {
            tenant: "t2",
            host: "node-b-auto",
        }
        .input(),
    );
    let manual = case(&root, "t2", "node-b-manual");
    assert_eq!(auto.intent, Some(Intent::Permanent));
    assert_eq!(auto.commit_step, Some(9));
    assert_eq!(manual.commit_step, Some(10));
    assert_eq!(auto.reversible_through, 3);
    let p = auto.point_of_no_return.as_ref().unwrap();
    assert_eq!(
        (p.step, p.guard.as_deref(), p.cost.as_str(), p.ack.as_str()),
        (4, Some("fence_verified_off"), "fence_verdict", "none")
    );
    assert_eq!(
        manual.point_of_no_return.as_ref().map(|p| p.ack.as_str()),
        Some("humans()")
    );
    assert!(!auto.holds_at.is_empty());
    assert_eq!(auto.held_indefinitely, vec![5, 6, 8]);
    assert_eq!(manual.held_indefinitely, vec![6, 7, 9]);
    assert_eq!(auto.deferred, vec![8]);
    assert_eq!(manual.deferred, vec![9]);
    assert_eq!(
        auto.steps.iter().filter(|s| s.refusal == "knell").count(),
        1
    );
    assert_eq!(
        manual.steps.iter().filter(|s| s.refusal == "knell").count(),
        2
    );
    assert_eq!(
        ir_auto.plan.strictness,
        Strictness::Strict,
        "the per-guest loop checks clean under strict"
    );
    assert!(auto.may_conflicts.is_empty());
    assert_eq!(auto.mode, "auto");
    assert_eq!(manual.mode, "manual");
    assert_eq!(auto.gate, None, "the auto plan contains no human wait");
    assert!(auto.steps.iter().all(|s| s.gate.is_none()));
    assert_eq!(auto.induced_defer, vec![5, 6, 8]);
}

#[test]
fn t3_commit_confirmed_change_section_8_3() {
    let root = repo_root().unwrap();
    let pf = case(&root, "t3", "fw-01");
    let win = case(&root, "t3", "fw-win-01");
    for v in [&pf, &win] {
        assert_eq!(v.intent, Some(Intent::Permanent));
        assert_eq!(v.commit_step, Some(4));
        assert_eq!(v.reversible_through, 1);
        let b = v.backstop.as_ref().unwrap();
        assert_eq!(b.covers, vec![1]);
        assert_eq!(b.installed_before, Some(1));
        assert_eq!(b.armed_before, Some(1));
        assert_eq!(b.armed_after, None);
    }
    let conditionals = |v: &Verdict| {
        v.steps
            .iter()
            .map(|s| s.conditional.clone())
            .collect::<Vec<_>>()
    };
    assert_eq!(
        conditionals(&pf),
        vec![
            Some("foreign region in file:/etc/pf.conf".into()),
            None,
            None,
            None
        ]
    );
    assert_eq!(conditionals(&win), vec![None, None, None, None]);
}

#[test]
fn t4_reactive_host_section_8_4() {
    let root = repo_root().unwrap();
    let clobber = case(&root, "t4", "site-ctl");
    let defer = case(&root, "t4", "site-ctl-defer");
    for v in [&clobber, &defer] {
        assert_eq!(v.intent, Some(Intent::Temporary));
        assert_eq!(v.wane, Some(Duration::new(7200)));
        assert_eq!(v.reversible_through, 1);
        assert_eq!(v.backstop, None);
    }
    assert_eq!(
        clobber
            .steps
            .iter()
            .map(|s| s.drift.clone())
            .collect::<Vec<_>>(),
        vec![Some("clobber".into())]
    );
    assert_eq!(
        defer
            .steps
            .iter()
            .map(|s| s.drift.clone())
            .collect::<Vec<_>>(),
        vec![Some("defer".into())]
    );
    assert_eq!(
        clobber
            .hosts_touched
            .iter()
            .find(|(n, _)| *n == 1)
            .map(|(_, h)| h.clone()),
        Some(vec![HostTouched::Host {
            host: "site-ctl".into(),
            directory: "controller".into()
        }])
    );
    assert_eq!(clobber.controller_only_undos, vec![1]);
}
